[editor][client] render array objects in settings properly

Ankush-lastmile opened the pull request:
[editor][client] render array objects in settings properly





Upon loading the OpenAI function calling aiconfig, the editor client does not render the function data. This is because the functions are an array, and the client does not initialize array data.

In this diff, if the Settings data is an array with an initial value, properly set/intiailize and render the array data (inside SettingsPropertyRenderer)

Also updated the object renderer/settings to render json editor for an unspecified Json

## Testplan

Load the function calling aiconfig and open the model settings drawer. Notice the functions get rendered.

Modified function call information inside the json editor. Tested the save was successful.

https://github.com/lastmile-ai/aiconfig/assets/141073967/f477f970-3de1-4c7b-adb4-eba1a8ca0bce

rholinshead commented on python/src/aiconfig/editor/client/src/components/SettingsPropertyRenderer.tsx:
```diff
@@ -124,6 +125,33 @@ export default function SettingsPropertyRenderer({
 
+ 
+
+  useEffect(() => {
```
We don't need an effect for this. This is just setting the initial state for itemControls so we should instead just initialize it inline above using the init function for useState:

```
const [itemControls, setItemControls] = useState<JSX.Element[]>(() => {
  if (!Array.isArray(initialValue)) {
    return [];
  }

  return initialValue.map((arrayItem,index) => {
    const key = uniqueId();
    return (
      <Group key={key}>
        <SettingsPropertyRenderer
          propertyName=""
          property={property.items}
          initialValue={arrayItem}
          setValue={newItem => {
            itemValues.current.set(key, newItem);
            setAndPropagateValue(Array.from(itemValues.current.values()));
          }}
       />
       <ActionIcon onClick={() => removeItemFromList(key)}>
         <IconTrash size={16} />
       </ActionIcon>
     </Group>
    );
  });
});
```

Ankush-lastmile replied:
updated as suggested. Is there anything bad that would happen if we were to continue to use `useEffect()`?

rholinshead replied:
Nothing critically bad, it's just the scenario I told you about --> doing it in a useEffect would first render with default [] in state, then hit the effect and then re-render again with new state from the effect. It's an extra full render cycle. This way the state is set initially before anything is ever added to the DOM and no re-render is needed

rholinshead commented on python/src/aiconfig/editor/client/src/components/SettingsPropertyRenderer.tsx:
```diff
@@ -367,8 +395,14 @@ export default function SettingsPropertyRenderer({
         );
+      } else {
+        propertyControl = (
+          <JSONEditor
+            content={initialValue as JSONObject}
+            onChangeContent={setAndPropagateValue}
+          />
```
```suggestion
        <Stack>
             <PropertyLabel
                propertyName={propertyName}
                propertyDescription={propertyDescription}
              />
            <JSONEditor
            content={initialValue as JSONObject}
            onChangeContent={setAndPropagateValue}
          />
        </Stack>
```

Let's add the label

rholinshead replied:
Also can you play around with it to see if we can get the width to look nicer? Maybe wrap the editor here with a div with 100% width or something?

Ankush-lastmile replied:
added label. Tried setting width to 100%, that did nothing. Anything > 100 would make it render nice but it would throw an exception on render

rholinshead replied:
Ya we won't want to go over 100% because there will be logic happening in monaco editor internals to handle resizing, probably based on parent element or something like that, and exceeding the parent width will cause troubles there

rholinshead commented on python/src/aiconfig/editor/client/src/components/SettingsPropertyRenderer.tsx:
```diff
@@ -90,7 +91,7 @@ export default function SettingsPropertyRenderer({
   // TODO: Should initialize with values from settings if available
   const [itemControls, setItemControls] = useState<JSX.Element[]>([]);
-  const itemValues = useRef(new Map<string, JSONValue>());
+  const itemValues = useRef(propertyValue ?? new Map<string, JSONValue>());
```
We shouldn't use propertyValue here since that's a state object that shouldn't be mutated directly (need to use setState for that). Instead, we should construct this map with the propertyValue. Maybe something like:

```
const itemValues = useRef(
  Array.isArray(propertyValue) ? new Map(propertyValue.map(val => [uniqueId(), val])) : 
  new Map<string, JSONValue>());
```

Ankush-lastmile replied:
Updated. 

>that shouldn't be mutated directly (need to use setState for that

Makes sense, but then why is it okay to set it to a newMap with the same property value?

rholinshead replied:
The code never mutates the map values, just the map itself. Before it would have been modifying the actual propertyValue object

rholinshead commented on python/src/aiconfig/editor/client/src/components/SettingsPropertyRenderer.tsx:
```diff
@@ -367,8 +398,22 @@ export default function SettingsPropertyRenderer({
         );
+      } else {
+        propertyControl = (
+          <div style={{width: "100%"}}>
+          <Stack>
+            <PropertyLabel
+             propertyName={propertyName}
+             propertyDescription={propertyDescription}
+           />
+            <JSONEditor
+              content={initialValue as JSONObject}
+              onChangeContent={setAndPropagateValue}
+            />
+          </Stack>
+          </div>
```
```suggestion
          <Stack>
            <PropertyLabel
             propertyName={propertyName}
             propertyDescription={propertyDescription}
           />
           <div style={{width: "100%"}}>
              <JSONEditor
                content={initialValue as JSONObject}
                onChangeContent={setAndPropagateValue}
              />
            </div>
          </Stack>
```
Let's wrap the div around the JSONEditor only. If that still doesn't make it wider can you try {{minWidth: "350px}} or something like that? We have a min width of 400px for the settings container (https://github.com/lastmile-ai/aiconfig/blob/3b91a530a376faf13b2ae7ad3bf0ba1b20114548/python/src/aiconfig/editor/client/src/components/prompt/PromptActionBar.tsx#L48)

rholinshead commented on python/src/aiconfig/editor/client/src/components/SettingsPropertyRenderer.tsx:
```diff
@@ -13,13 +13,14 @@ import {
   Select,
 } from "@mantine/core";
-import { useState, useCallback, useRef } from "react";
+import { useState, useCallback, useRef, useEffect } from "react";
```
```suggestion
import { useState, useCallback, useRef } from "react";
```
not needed

Ankush-lastmile replied:
removed

rholinshead reviewed (approved):
Accepting to unblock, but can you also test that changing a setting and saving the config properly persists these generic json settings?

And before landing let's try and get that width sorted out. Please ping me if you can't get it and I can play around with it a bit as well

Ankush-lastmile commented:
rebase, updated json editor for model settings to be 350px wide.

Unfortunately getting a weird side effect where updating the json editor inside settings now adds to it than updating.

https://github.com/lastmile-ai/aiconfig/assets/141073967/47cf6275-711a-4ba7-a2fd-6a24ab48fb2f

Ankush-lastmile commented:
fixed some spacing

Ankush-lastmile commented:
>Unfortunately getting a weird side effect where updating the json editor inside settings now adds to it than updating.

Fixed this by
1. Initialize `itemValues` first with propertyValue
2. Then Initialize `itemControls` with the resultant `itemValues`

This way the JSX keys lineup. 

Also removed the TODO comment

rholinshead reviewed (approved):
<img src="https://media3.giphy.com/media/3oEjHWXddcCOGZNmFO/giphy.gif"/>
